[doc] Add tutorial about reusable tests (#1895)
eth-cscs/reframe: docs/tutorial_advanced.rst

For a complete list of the available attributes of a specific container platform, please have a look at the :ref:`container-platforms` section of the :doc:`regression_test_api` guide.
On how to configure ReFrame for running containerized tests, please have a look at the :ref:`container-platform-configuration` section of the :doc:`config_reference`.

Writing reusable tests
----------------------

> .. versionadded:: 3.5

Reviewer: ```suggestion
.. versionadded:: 3.5.0
```

So far, all the examples shown above were tight to a particular system or configuration, which makes reusing these tests in other systems not straightforward.
However, the introduction of the :py:func:`~reframe.core.pipeline.RegressionTest.parameter` and :py:func:`~reframe.core.pipeline.RegressionTest.variable` ReFrame built-ins solves this problem, eliminating the need to specify any of the test variables in the :func:`__init__` method.
Hence, these parameters and variables can be treated as simple class attributes, which allows us to leverage Python's class inheritance and write more modular tests.
For simplicity, we illustrate this concept with the above :class:`ContainerTest` example, where the goal here is to re-write this test as a library that users can simply import from and derive their tests without having to rewrite the bulk of the test.

> Also, for illustrative purposes, we parametrize this library test on a few different images (the above example just used ``ubuntu:18.04``) and throw the container commands into a separate bash script just to create some source files.

Reviewer 2: ```suggestion
Also, for illustrative purposes, we parameterize this library test on a few different image tags (the above example just used ``ubuntu:18.04``) and throw the container commands into a separate bash script just to create some source files.
```

Author: Taking sides on the `parametrize` vs `parameterize` uh? 😂

   cat tutorials/advanced/library/lib/__init__.py

.. literalinclude:: ../tutorials/advanced/library/lib/__init__.py
   :lines: 6-
   :emphasize-lines: 8-17

Note that the class :class:`ContainerBase` is not decorated since it does not specify the required variables ``valid_systems`` and ``valid_prog_environs``, and it declares the ``platform`` parameter without any defined values assigned.
Hence, the user can simply derive from this test and specialize it to use the desired container platforms.
Since the parameters are defined directly in the class body, the user is also free to override or extend any of the other parameters in a derived test.
In this example, we have parametrized the base test to run with the ``ubuntu:18.04`` and ``ubuntu:20.04`` images, but these values from ``dist`` (and also the ``dist_name`` variable) could be modified by the derived class if needed.

On the other hand, the rest of the test depends on the values from the test parameters, and a parameter is only assigned a specific value after the class has been instantiated.
Thus, the rest of the test is expressed as hooks, without the need to write anything in the :func:`__init__` method.

> In fact, note that writing the test in this way permits having hooks that depend on undefined variables or parameters.

Reviewer 2: ```suggestion
In fact, writing the test in this way permits having hooks that depend on undefined variables or parameters.
```

This is the case with the :func:`set_container_platform` hook, which depends on the undefined parameter ``platform``.

> Hence, the derived test **must** define all the required parameters and variables; otherwise ReFrame will notice that the test is not well defined and it will raise an error accordingly.

Reviewer 2: ```suggestion
Hence, the derived test **must** define all the required parameters and variables; otherwise ReFrame will notice that the test is not well defined and will raise an error accordingly.
```

Before moving onwards to the derived test, note that the :class:`ContainerBase` class takes the additional argument ``pin_prefix=True``, which locks the prefix of all derived tests to this base test.
This will allow the retrieval of the sources located in the library by any derived test, regardless of what their containing directory is.

.. code-block:: console

   cat tutorials/advanced/library/lib/src/get_os_release.sh

.. literalinclude:: ../tutorials/advanced/library/lib/src/get_os_release.sh
   :lines: 1-

> Now from the user's perspective, the only thing to do is importing the above base test and specify the required variables and parameters.

Reviewer 2: ```suggestion
Now from the user's perspective, the only thing to do is import the above base test and specify the required variables and parameters.
```

Author: `import` is not correct here. It's either `importing` or `to import`. Changed it to the latter.
